feat(redis): basic redis support

mrsimb opened the pull request:
As per title, added basic redis support. Resolves #80.

Stuff that works:
- listing keys, their types, ttl, and occupied memory
- filtering keys (use `raw filter`)
- querying (similar to redis-cli, thanks to the [redis-splitargs](https://github.com/luin/splitargs) author for this)
- listing redis server info
- pagination (though it needs cursor pagination, which is only supported when exporting data in beekeeper)
- selecting database
- redis connection form
- showing/editing key content in table view
- showing/editing key content in JSON viewer
- syntax highlighting and completions in query editor
- changing key name
- changing key TTL in seconds
- streams

Stuff that doesn't (yet):
- showing pub/subs (not sure if i want to add it)

mrsimb commented:
This might be expanded to showing graphs/plots (if applicable) when selecting multiple rows, but i'm not sure if there will be demand on such feature, since you can do the same by copying the data to excel sheet.

mrsimb commented:
However you can take it from here if you'd like, i definitely wouldn't mind :)

not-night-but commented on apps/studio/src/lib/db/clients/redis.ts:
```diff
@@ -0,0 +1,634 @@
+    throw new Error("Not implemented");
+  }
+
+  async rawExecuteQuery(
+    _query: string
+  ): Promise<never> {
+    throw new Error("Where is this called from?");
+  }
```
Raw execute query is called from the BasicDatabaseClient. Typically what we like to do is just implement this, and then when we want to query from within the driver, we call `this.driverExecuteSingle` or `this.driverExecuteMultiple` (execute a single query or execute multiple). It provides guards for readonly mode and does some other things. I'm not sure how relevent this will be for redis, as I'm honestly not too familiar with the db. We did have to completely cut this out of our mongo driver for example, because it just didn't fit that paradigm.

mrsimb replied:
Is this related to ability to select some code in query editor to run it, isolated from the whole code?
Read only mode would be a nice feature to have for redis.

not-night-but commented on apps/studio/src/shared/lib/dialects/redis.ts:
```diff
@@ -0,0 +1,74 @@
+export const RedisData: DialectData = {
+  columnTypes: [],
+  usesOffsetPagination: true,
+  queryDialectOverride: 'generic', // TODO: idk what to put there
+  textEditorMode: 'text/x-pgsql', // TODO: there's no "redis" mode for codemirror
```
The dialect override I think was specifically used for mongo, but I don't think it's relevant here. In the case of mongo, we have the ability to query with a postgres-like syntax, which is why we override the query dialect for it (because otherwise it just uses the connection type). We may need to think about looking into developing a redis mode for codemirror

mrsimb replied:
Agree about codemirror mode

not-night-but commented on apps/studio/src/shared/lib/dialects/redis.ts:
```diff
@@ -0,0 +1,74 @@
+  textEditorMode: 'text/x-pgsql', // TODO: there's no "redis" mode for codemirror
+  disabledFeatures: {
+    // rawFilters: true, // TODO: how to disable non-raw filters?
+    shell: true,
```
Does redis need a shell like experience? Currently our shell is pretty heavily tied to mongo, but we could definitely make it more generic if this is the case.

not-night-but replied:
![image](https://github.com/user-attachments/assets/062cc9b2-9dc8-4318-ab82-9ec96c69a32f)

This is shell. If you're wanting redis to have a REPL shell like this, then we'll need to update the shell to be less tied to mongo (I can help with this).

mrsimb replied:
oh now i get it
no, redis has no shell feature

mrsimb commented on apps/studio/src/lib/db/clients/redis.ts:
```diff
@@ -0,0 +1,634 @@
+
+  async getCompletions(): Promise<string[]> {
+    return [
+      "GET",
+      "SET",
+      "DEL",
+      "EXISTS",
+      "EXPIRE",
+      "TTL",
+      "TYPE",
+      "HGET",
+      "HSET",
+      "HDEL",
+      "HKEYS",
+      "HVALS",
+      "HGETALL",
+      "LPUSH",
+      "RPUSH",
+      "LPOP",
+      "RPOP",
+      "LLEN",
+      "LRANGE",
+      "SADD",
+      "SREM",
+      "SMEMBERS",
+      "SCARD",
+      "ZADD",
+      "ZREM",
+      "ZRANGE",
+      "ZCARD",
+      "ZSCORE",
+      "KEYS",
+      "SCAN",
+      "FLUSHDB",
+      "FLUSHALL",
+      "INFO",
+      "PING",
+    ];
+  }
+
```
I thought this is required for auto completion to work. If custom redis codemirror plugin would do this, then this is not needed. I'll remove this. In any case, making a plugin for redis is a bit out of scope for me now, maybe after the basic stuff is finished.

not-night-but replied:
For sure! I think basic functionality is most important before we start getting in the weeds of stuff.

This function was only added for the mongo shell integration, but depending on whether or not this is a shell, we may want to keep this implementation.

mrsimb commented:
@not-night-but  I added value loading and displaying for these types in the JSON sidebar: "string" | "list" | "hash" | "set" | "zset" | "ReJSON-RL" (JSON addon).

You set some values in query editor to view them in "keys" table. See how to use redis-cli for help or generate some queries in chatgpt.

I originally planned to create a separate, generalized key-value storage view alongside the JSON viewer, but I ditched that idea since it’s too time consuming.

I’m going to have to leave this as is for now, as I have a pretty busy period at work. I’ll keep monitoring this thread though, so feel free to ask me anything!

mrsimb commented:
<img width="602" height="380" alt="image" src="https://github.com/user-attachments/assets/30d201f0-9ce3-41d1-864f-7e25810e0a42" />
redis autocomplete ✨

mrsimb commented:
<img width="973" height="720" alt="image" src="https://github.com/user-attachments/assets/66924407-7f1b-4b18-8dfd-ef1717013a78" />

value editor sidebar (i ditched idea with using json viewer for editing values, it is not convenient)

mrsimb commented:
@not-night-but can you review this please so i know if our opinions still align?

not-night-but commented on apps/studio/src/lib/db/clients/redis.ts:
```diff
@@ -0,0 +1,853 @@
+    return `SCAN 0 MATCH * COUNT ${limit}`;
+  }
+
+  async selectTopStream(): Promise<StreamResults> {
+    throw new Error("Not implemented");
+  }
+
+  async queryStream(): Promise<StreamResults> {
+    throw new Error("Not implemented");
+  }
```
Not sure if this is something redis users would want, but looking at the scan code you have, we could probably implement these relatively easy which would allow for 'table' export and 'query' export functionality

not-night-but commented on apps/studio/src/lib/db/clients/redis.ts:
```diff
@@ -0,0 +1,853 @@
+  async executeApplyChanges(
+    changes: TableChanges
+  ): Promise<TableUpdateResult[]> {
+    log.debug("Redis executeApplyChanges called with:", changes);
```
Ignorant non-redis user question: are transactions possible, or anything approaching a transaction? Generally we try to wrap everything in this function in a transaction so we can roll back if one of the changes fails

not-night-but commented on apps/studio/vite.config.mjs:
```diff
@@ -16,8 +16,16 @@ export default defineConfig({
-    exclude: []
-  },
+    exclude: [
+      // Exclude native modules from optimization
+      // Without this, the build fails :(
```
Huh. That's a new one

mrsimb replied:
Yeah, i've read somewhere in the code that someone wanted to get rid of those in client runtime, so that's how it is done.

not-night-but reviewed (commented):
These are just some general thoughts, but this is an insane contribution! Thank you so much for your work on this (especially the codemirror stuff, I know that can be a bloody nightmare). I'm not sure where you consider your work on this to be, but I'm available for questions/discussions if you need.

mrsimb commented:
Added "value" column. I hesitated to do this at first, because there can be massive amounts of data in keys content. But beekeeper might just handle it. Also, maybe using json viewer isn't that bad...

<img width="1439" height="876" alt="image" src="https://github.com/user-attachments/assets/f7c52f90-7b22-401e-9d81-37f4c4489aad" />

mrsimb commented on apps/studio/src/components/ConnectionInterface.vue:
```diff
@@ -83,6 +83,11 @@
                   :testing="testing"
                 />
+                <redis-form
+                  v-else-if="config.connectionType === 'redis' && isUltimate"
```
Made it available for non ultimate users, you guys decide which should it be later.

mrsimb commented:
@not-night-but I addressed most of the stuff you wrote, and reverted to using json-viewer for editing values. The implementation is almost complete, only value setting for complex types and some minor issues are left i think.

mrsimb commented:
@not-night-but Good to hear (read?) that :)

I didn't know there's modal editor, it should probably be invoked by default when user tries to edit a redis value. Other than that, it suits perfectly. Just found a bug though, both when editing using the modal and when using the in-place cell input. I'll fix it tomorrow.

And yeah, i'll let you guys know if i have any new ideas for features.

mrsimb commented on apps/studio/src/lib/db/clients/redis.ts:
```diff
@@ -0,0 +1,853 @@
+    const results: NgQueryResult[] = [];
+
+    for (const command of commands) {
+      const args = splitargs(command);
+      try {
+        const result = await this.redis.call.call(this.redis, ...args);
```
Oh, it actually supports evaluating lua code https://redis.io/docs/latest/develop/programmability/eval-intro/

mrsimb commented:
@not-night-but I fixed issue when editing values in TableTable's cell or in modal. Also added proper validation for input, so its harder to f*** up your data. Oh and it reads real database count now.

The branch is rebased to the current master (had to create new file for migration, since dev-9 was occupied by Trino).

not-night-but commented on apps/studio/src/shared/lib/dialects/models.ts:
```diff
@@ -161,8 +162,7 @@ export interface DialectData {
     duplicateTable?: boolean,
     export?: {
-      sql?: boolean,
-      stream?: boolean
```
We also still need this

not-night-but commented on apps/studio/src/lib/db/clients/redis.ts:
```diff
@@ -0,0 +1,875 @@
+  };
+}
+
+class RedisChangeBuilder extends ChangeBuilderBase {
+  dialect: Dialect = "postgresql";
+
+  wrapIdentifier(value: string): string {
+    return value;
+  }
+
+  wrapLiteral(value: string): string {
+    return value;
+  }
+
+  escapeString(value: string): string {
+    return value;
+  }
+
+  alterType(_column: string, _newType: string): string {
+    throw new Error("Not supported");
+  }
+
+  alterDefault(_column: string, _newDefault: string | null): string {
+    throw new Error("Not supported");
+  }
+
+  alterNullable(_column: string, _nullable: boolean): string {
+    throw new Error("Not supported");
+  }
+
+  addColumn(_item: SchemaItem): string {
+    throw new Error("Not supported");
+  }
+
+  dropColumn(_column: string): string {
+    throw new Error("Not supported");
+  }
+}
```
I get this is kinda a throwaway class, but I'd still put it in it's own file. (`apps/studio/src/shared/lib/sql/change_builder/`)

mrsimb replied:
Moved

not-night-but reviewed (dismissed):
Okay just a couple little things. And it looks like possibly some merge issues? Looks like some trino stuff has been removed.

mrsimb commented:
@not-night-but yeah it was after rebase I guess. Do you mind if i just reset the branch to a state before the rebase, and you'll handle merge conflicts?

not-night-but commented:
@mrsimb yeah reading can be super annoying lol. Go for it

mrsimb commented on apps/studio/src/components/connection/RedisForm.vue:
```diff
@@ -0,0 +1,101 @@
+      <small class="form-text text-muted">
+        Not applicable to Redis Cluster
+      </small>
+    </div>
```
Changed the input type

mrsimb commented:
@not-night-but Done, you can check it out now
